### PR TITLE
ipsec: T1925: Fixed `show vpn ipsec sa` output

### PR DESCRIPTION
**Pull request by zdc**
## Change Summary

Fixed `show vpn ipsec sa` output

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T1925

## Component(s) name

ipsec

## Proposed changes

After the a1aaf4fb9c0e4111670ef3dd491796fa35a2311f commit, only single (latest) CHILD_SA for each connection can be displayed in the `show vpn ipsec sa` output. This commit backs the proper behavior for the command and adds a little optimization to the formatter to make it easier.

## How to test

One of the ways how to test is to configure a DMVPN connectivity with multiple peers. In the output, you will see only the latest peer or even a parent SA if the latest peer in a list is not connected, what is wrong. After this change, the output will contain all CHILD_SA.

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by sever-sever** [comment]
This breaks output for the ipsec, the difference with current and new output (new_sa.py from your repo):
```
root@r11-roll:/home/vyos# show vpn ipsec sa
Connection                  State    Uptime    Bytes In/Out    Packets In/Out    Remote address    Remote ID    Proposal
--------------------------  -------  --------  --------------  ----------------  ----------------  -----------  ------------------------
peer_192-0-2-14_tunnel_0    up       1m55s     0B/0B           0/0               192.0.2.14        N/A          AES_GCM_16_256/MODP_2048
peer_192-0-2-14_tunnel_1    down     N/A       N/A             N/A               N/A               N/A          N/A
peer_192-0-2-14_tunnel_2    down     N/A       N/A             N/A               N/A               N/A          N/A
peer_192-0-2-14_tunnel_3    down     N/A       N/A             N/A               N/A               N/A          N/A
peer_203-0-113-14_tunnel_0  up       2m16s     0B/0B           0/0               203.0.113.14      N/A          AES_GCM_16_256/MODP_2048
root@r11-roll:/home/vyos# 
root@r11-roll:/home/vyos# 
root@r11-roll:/home/vyos# ./new_sa.py 
Connection                  State    Uptime    Bytes In/Out    Packets In/Out    Remote address    Remote ID     Proposal
--------------------------  -------  --------  --------------  ----------------  ----------------  ------------  ------------------------
peer_192-0-2-14_tunnel_0    up       1m57s     0B/0B           0B/0B             192.0.2.14        192.0.2.14    AES_GCM_16_256/MODP_2048
peer_203-0-113-14_tunnel_0  up       2m18s     0B/0B           0B/0B             203.0.113.14      203.0.113.14  AES_GCM_16_256/MODP_2048
root@r11-roll:/home/vyos# 
```
In the tested configuration **only tunnel0 are active**/up/established/installed:
```
set vpn ipsec esp-group grp-ESP compression 'disable'
set vpn ipsec esp-group grp-ESP lifetime '28800'
set vpn ipsec esp-group grp-ESP mode 'tunnel'
set vpn ipsec esp-group grp-ESP pfs 'dh-group14'
set vpn ipsec esp-group grp-ESP proposal 10 encryption 'aes256gcm128'
set vpn ipsec esp-group grp-ESP proposal 10 hash 'sha256'
set vpn ipsec ike-group grp-IKE dead-peer-detection action 'hold'
set vpn ipsec ike-group grp-IKE dead-peer-detection interval '30'
set vpn ipsec ike-group grp-IKE dead-peer-detection timeout '120'
set vpn ipsec ike-group grp-IKE ikev2-reauth 'no'
set vpn ipsec ike-group grp-IKE key-exchange 'ikev2'
set vpn ipsec ike-group grp-IKE lifetime '86400'
set vpn ipsec ike-group grp-IKE mobike 'disable'
set vpn ipsec ike-group grp-IKE proposal 10 dh-group '14'
set vpn ipsec ike-group grp-IKE proposal 10 encryption 'aes256gcm128'
set vpn ipsec ike-group grp-IKE proposal 10 hash 'sha256'
set vpn ipsec interface 'eth1'
set vpn ipsec site-to-site peer 192.0.2.14 authentication id '192.0.2.1'
set vpn ipsec site-to-site peer 192.0.2.14 authentication mode 'pre-shared-secret'
set vpn ipsec site-to-site peer 192.0.2.14 authentication pre-shared-secret 'SSSeeccRetT'
set vpn ipsec site-to-site peer 192.0.2.14 authentication remote-id '192.0.2.14'
set vpn ipsec site-to-site peer 192.0.2.14 connection-type 'initiate'
set vpn ipsec site-to-site peer 192.0.2.14 ike-group 'grp-IKE'
set vpn ipsec site-to-site peer 192.0.2.14 ikev2-reauth 'inherit'
set vpn ipsec site-to-site peer 192.0.2.14 local-address '192.0.2.1'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 0 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 0 local prefix '172.16.0.0/24'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 0 remote prefix '10.0.0.0/24'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 1 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 1 local prefix '172.16.0.0/24'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 1 remote prefix '10.0.11.0/24'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 2 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 2 local prefix '172.16.0.0/24'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 2 remote prefix '10.0.22.0/24'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 3 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 3 local prefix '172.16.0.0/24'
set vpn ipsec site-to-site peer 192.0.2.14 tunnel 3 remote prefix '10.0.33.0/24'
set vpn ipsec site-to-site peer 203.0.113.14 authentication id '203.0.113.1'
set vpn ipsec site-to-site peer 203.0.113.14 authentication mode 'pre-shared-secret'
set vpn ipsec site-to-site peer 203.0.113.14 authentication pre-shared-secret 'SSSeeccRetT'
set vpn ipsec site-to-site peer 203.0.113.14 authentication remote-id '203.0.113.14'
set vpn ipsec site-to-site peer 203.0.113.14 connection-type 'initiate'
set vpn ipsec site-to-site peer 203.0.113.14 ike-group 'grp-IKE'
set vpn ipsec site-to-site peer 203.0.113.14 ikev2-reauth 'inherit'
set vpn ipsec site-to-site peer 203.0.113.14 local-address '203.0.113.1'
set vpn ipsec site-to-site peer 203.0.113.14 tunnel 0 esp-group 'grp-ESP'
set vpn ipsec site-to-site peer 203.0.113.14 tunnel 0 local prefix '172.16.0.0/24'
set vpn ipsec site-to-site peer 203.0.113.14 tunnel 0 remote prefix '10.0.0.0/24'

```

File: src/op_mode/show_ipsec_sa.py [hunk 1]
```diff
@@ -14,119 +14,115 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-import re
-import sys
+from hurry import filesize
+from re import split as re_split
+from sys import exit as sys_exit
```
Comment (c-po) [inline src/op_mode/show_ipsec_sa.py]: Please simply import it as exit to match all other files.

**Review by c-po — CHANGES_REQUESTED** [review]

File: src/op_mode/show_ipsec_sa.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 #
-# Copyright (C) 2019 VyOS maintainers and contributors
+# Copyright (C) 2019-2022 VyOS maintainers and contributors
```
Comment (dmbaturin) [inline src/op_mode/show_ipsec_sa.py]: Nitpick: could be just 2022.

**Review by dmbaturin — APPROVED** [review]
Looks good other than the nitpick.

**Comment by c-po** [comment]
@sever-sever when I read your comment, does it mean this PR is faulty?

**Comment by dmbaturin** [comment]
@c-po Which exact comment?

**Comment by sever-sever** [comment]
> @sever-sever when I read your comment, does it mean this PR is faulty?

It depends on what we want to see in `show vpn ipsec sa`
With new implementation we won't see child SA's with not `installed` state, i.e. we won't see all tunnels in `down` state
On another hand child SA's != tunnel

In 1.1.8, 1.2.x we see all down/up. Why I ask it, users use custom scripts that that parse "show vpn ipsec sa" detect tunnels in "down" state and if they detect it they do some action like "restart vpn/restart ipsec/re-load configuration" so on.

If we want to display just established/installed SA's, we should describe it in the documentation that behavior was changed and you will see just SA's which was established/installed.
Somehow is it possible I prefer to be able to see all states `up/down` in our case.

**Comment by sever-sever** [comment]
> @c-po Which exact comment?

https://github.com/vyos/vyos-1x/pull/1133#issuecomment-1004846049

**Comment by c-po** [comment]
Maybe add an option to also dee down tunnels?

I myself used such scripts tonrestart down tunnels.

**Review by dmbaturin — CHANGES_REQUESTED** [review]
I don't think we can merge this until the down tunnels can be displayed.

**Comment by zdc** [comment]
@dmbaturin Adding tunnels in a "down" state is an easy task. But I cannot agree that this is necessary here. Let me share my opinion.

First what we need to understand are two things:
- one tunnel/VTI may have more than one SA;
- there can be no SA for a tunnel/VTI at all.

Another side of the case is that we actually never had a proper command to display tunnels/VTI states. And `show vpn ipsec sa` just was the closest to this unexisting command. Which, in the end, ruined its original purpose. Now, instead of displaying SAs, it displays some weird mix of configured tunnels, services like L2TP/DMVPN, SAs, and calculated states that have nothing to SAs.
After migration to `vici` that you did in 2019 (https://github.com/vyos/vyos-1x/commit/852253ce2c5cae371663d0f15b9f416b0848e9b9#diff-951ed3f63d1f64e10b581fe8102d5c306cd80de9b4e7349a4d7e9a422dfa87c8) this command worked as it should be, but in the middle of 2021 it was modified, and wrong behavior appeared again (https://github.com/vyos/vyos-1x/commit/a1aaf4fb9c0e4111670ef3dd491796fa35a2311f#diff-951ed3f63d1f64e10b581fe8102d5c306cd80de9b4e7349a4d7e9a422dfa87c8).

This is a good example of a wrong understanding of the original command purpose:
> With new implementation we won't see child SA's with not installed state

Actually, we will see them as "down". What we will not see - SAs that does not exist. I hope that this is obvious that when you ask "show me SAs" you expect to see real SAs, without fictional ones.

So, we should realize that now we are trying to solve a problem with displaying states of tunnels with the wrong tool, which makes this tool unusable for its original purpose.

Because of this, I would suggest two things:

- leave `show vpn ipsec sa` with the functionality that is intended to it - display SAs;
- introduce another command, possibly `show vpn ipsec connections`, maybe even with compatible output (same fields), that will add missing features and display information that you expect to see about tunnels/VTI/DMVPN/L2TP statuses.

Because mixing them both in a single output will necessarily make it unusable for one of the goals.

@dmbaturin If I was not convincing enough and you still would like to add fake SAs to the output of this command, I will do this. :)

File: src/op_mode/show_ipsec_sa.py (hunk 1 above)
Comment (zdc) [reply src/op_mode/show_ipsec_sa.py]: Thanks! Fixed. Also, regrouped imports according to the PEP8.

File: src/op_mode/show_ipsec_sa.py (hunk 2 above)
Comment (zdc) [reply src/op_mode/show_ipsec_sa.py]: Got it. Done.

**Pull request merged**